xclim/sdba/adjustment.py

```diff
@@ -406,6 +459,168 @@ def _adjust(self, sim, interp="nearest", extrapolation="constant"):
+        Similarly with `sim`. The cdf of the extreme values of `sim` is computed in reference
+        to the distribution fitted on `sim` and then the corresponding values (quantile / ppf)
+        in reference to the distribution fitted on `ref` are taken as the new bias-adjusted values.
+
+        Once new extreme values are found, a mixture from the original scen and corrected scen
+        is used in the result. For each original value :math:`S_i` and corrected value :math:`C_i`
+        the final extreme value :math:`V_i` is:
+
+        .. math::
+
+            V_i = C_i * \tau + S_i * (1 - \tau)
+
+        Where :math:`\tau` is a function of sim's extreme values :math:`F` and of arguments
+        ``frac`` (:math:`f`) and ``power`` (:math:`p`):
+
+        .. math::
+
+            \tau = \left(\frac{1}{f}\frac{S - min(S)}{max(S) - min(S)}\right)^p
+
+        References
+        ----------
+        .. [ClimateTools] https://juliaclimate.github.io/ClimateTools.jl/stable/
+        .. [RRJF2021] Roy, P., Rondeau-Genesse, G., Jalbert, J., Fournier, É. 2021. Climate Scenarios of Extreme Precipitation Using a Combination of Parametric and Non-Parametric Bias Correction Methods. Submitted to Climate Services, April 2021.
+        """
+        super().__init__(q_thresh=q_thresh, cluster_thresh=cluster_thresh, dist=dist)
+
+    def _train(self, ref, hist):
+        cluster_thresh = convert_units_to(self.cluster_thresh, ref)
+        hist = convert_units_to(hist, ref)
+
+        # Extreme value threshold computed relative to "large values".
+        # We use the mean between ref and hist here.
+        thresh = (
+            ref.where(ref >= cluster_thresh).quantile(self.q_thresh, dim="time")
+            + hist.where(hist >= cluster_thresh).quantile(self.q_thresh, dim="time")
+        ) / 2
+
+        # All large value clusters
+        ref_clusters = get_clusters(ref, thresh, cluster_thresh)
+        # Parameters of a genpareto (or other) distribution, we force te location at thresh.
+        fit_params = stats.fit(
+            ref_clusters.maximum - thresh, self.dist, dim="cluster", floc=0
+        )
+        # Param "loc" was fitted with 0, put thresh back
+        fit_params[..., 1] = fit_params[..., 1] + thresh
+
+        ds = xr.Dataset(
+            dict(fit_params=fit_params, thresh=thresh, nclusters=ref_clusters.nclusters)
+        )
+        ds.fit_params.attrs.update(
+            long_name=f"{self.dist} distribution parameters of ref",
+        )
+        ds.thresh.attrs.update(
+            long_name=f"{self.q_thresh * 100}th percentile extreme value threshold",
+            description=f"Mean of the {self.q_thresh * 100}th percentile of large values (x > {self.cluster_thresh}) of ref and hist.",
+        )
+        self.set_dataset(ds)
+
+    def _adjust(
+        self,
+        scen,
+        sim,
+        frac: float = 0.25,
+        power: float = 1.0,
+    ):
```

Comment:
Docstring. I know this is a private function, but it's not clear what power does.

Reply:
Right now the transition function is hardcoded and we played a bit with the `power` factor and settled on `1.0` in the end. What I'd like to do ultimately is provide the user with the tools so that they can provide their own definition of a transition function to be applied.

I guess that here we simply need a definition of the transition function somewhere in the docstrings. (I admit the function is a little convoluted).

Reply:
@huard, like in other adjustment object, everything is explained in the class docstring. The second part of the notes explains `frac` and `power`. In the same wy as we explain `interp` in EQM's docstring, not in `EQM._adjust`.